### PR TITLE
Add `TagHelperFactsService`.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Add API to enable the editor to query information on the state of `TagHelper`s within a Razor document.
- Refactored methods from `TagHelperDescriptorProvider` to be public static so our language service could use them.

#1120

### CODE REVIEW

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultRazorSyntaxFactsService.cs [hunk 1]
```diff
@@ -1,15 +1,15 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using Microsoft.AspNetCore.Razor.Evolution;
+using Microsoft.AspNetCore.Razor.Evolution.Legacy;
+using Microsoft.VisualStudio.Text;
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultRazorSyntaxFactsService.cs]: 👎

File: src/Microsoft.VisualStudio.LanguageServices.Razor/TagHelperFactsService.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Evolution;
+using Microsoft.AspNetCore.Razor.Evolution.Legacy;
+using System.Collections.Generic;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    public abstract class TagHelperFactsService
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/TagHelperFactsService.cs]: Let's put the `TagHelperDocumentContext` first.

**Comment by NTaylorMullen** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Razor.Evolution/TagMatchingRuleConventions.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+            if (rule == null)
+            {
+                throw new ArgumentNullException(nameof(rule));
+            }
+
+            if (!rule.Attributes.All(
+                requiredAttribute => tagAttributes.Any(
+                    attribute => requiredAttribute.IsMatch(attribute.Key, attribute.Value))))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/TagMatchingRuleConventions.cs]: /twitches 😲

Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Evolution/TagMatchingRuleConventions.cs]: We should get rid of `IsMatch` and put that here.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            {
+                throw new ArgumentNullException(nameof(binding));
+            }
+
+            var matchingBoundAttributes = new List<BoundAttributeDescriptor>();
+            foreach (var descriptor in binding.Descriptors)
+            {
+                foreach (var boundAttributeDescriptor in descriptor.BoundAttributes)
+                {
+                    if (boundAttributeDescriptor.CanMatchName(attributeName))
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs]: Same deal here, this should be on the conventions.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                foreach (var rule in descriptor.TagMatchingRules)
+                {
+                    if (TagMatchingRuleConventions.SatisfiesParentTag(parentTag, rule))
+                    {
+                        matchingDescriptors.Add(descriptor);
+                    }
+                }
+            }
+
+            var distinct = matchingDescriptors.Distinct(TagHelperDescriptorComparer.Default).ToList();
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs]: Why do we need this?

Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs]: We shouldn't need any of this silliness.

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Test
+{
+    public class DefaultTagHelperFactsServiceTest
+    {
+        // Purposefully not testing DefaultTagHelperFactsService.GetTagHelperBinding because it's a pass through
+        // into TagHelperDescriptorProvider.GetTagHelperBinding. That code is tested thoroughly.
```
Comment (rynowak) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs]: I would be nice to have just one basic test for it.

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,308 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Evolution;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Test
```
Comment (rynowak) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs]: no `.Test`

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs [hunk 8]
```diff
@@ -0,0 +1,308 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Razor.Evolution;
```
Comment (rynowak) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs]: Sort

File: test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+
+            // Act
+            var descriptors = service.GetTagHelpersGivenParent(documentContext, "div");
+
+            // Assert
+            Assert.Equal(expectedDescriptors, descriptors, TagHelperDescriptorComparer.CaseSensitive);
+        }
+
+        [Fact]
+        public void GetTagHelpersGivenParent_DoesNotReturnDuplicateDescriptors()
```
Comment (rynowak) [inline test/Microsoft.VisualStudio.LanguageServices.Razor.Test/DefaultTagHelperFactsServiceTest.cs]: I don't think we should do things like this. Ask a stupid question, get a stupid answer.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTagHelperFactsService.cs]: Total doh moment here. Will remove.

**Pull request merged**